src/stream/src/executor/row_merge.rs

```diff
@@ -0,0 +1,189 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use risingwave_common::bail;
+use risingwave_common::types::ToOwnedDatum;
+use risingwave_common::util::chunk_coalesce::DataChunkBuilder;
+use risingwave_common::util::column_index_mapping::ColIndexMapping;
+
+use super::barrier_align::*;
+use crate::executor::prelude::*;
+
+pub struct RowMergeExecutor {
+    ctx: ActorContextRef,
+    pub lhs_input: Executor,
+    pub rhs_input: Executor,
+    /// Maps input from the lhs to the output.
+    pub lhs_mapping: ColIndexMapping,
+    /// Maps input from the rhs to the output.
+    pub rhs_mapping: ColIndexMapping,
+    /// Output schema
+    pub schema: Schema,
+}
+
+impl RowMergeExecutor {
+    pub fn new(
+        ctx: ActorContextRef,
+        lhs_input: Executor,
+        rhs_input: Executor,
+        lhs_mapping: ColIndexMapping,
+        rhs_mapping: ColIndexMapping,
+        schema: Schema,
+    ) -> Self {
+        Self {
+            ctx,
+            lhs_input,
+            rhs_input,
+            lhs_mapping,
+            rhs_mapping,
+            schema,
+        }
+    }
+
+    #[try_stream(ok = Message, error = StreamExecutorError)]
+    pub async fn execute_inner(self) {
+        let lhs_mapping = self.lhs_mapping;
+        let rhs_mapping = self.rhs_mapping;
+        let data_types = self
+            .schema
+            .fields()
+            .iter()
+            .map(|f| f.data_type())
+            .collect::<Vec<_>>();
+
+        {
+            let mut lhs_buffer: Vec<StreamChunk> = Vec::with_capacity(1);
+            let mut rhs_buffer: Vec<StreamChunk> = Vec::with_capacity(1);
+            let aligned_stream = barrier_align(
+                self.lhs_input.execute(),
+                self.rhs_input.execute(),
+                self.ctx.id,
+                self.ctx.fragment_id,
+                self.ctx.streaming_metrics.clone(),
+                "Join",
```

Comment:
Should be RowMerge?

Reply:
Handled: https://github.com/risingwavelabs/risingwave/pull/17930/commits/ebf0545b60143632ff9e9cc574f3ef1f6e4047df